File: app/src/main/java/com/kunzisoft/keepass/database/action/node/CopyEntryRunnable.java
/*
 * Copyright 2018 Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.action.node;

import android.content.Context;
import android.util.Log;

import com.kunzisoft.keepass.database.Database;
import com.kunzisoft.keepass.database.PwEntry;
import com.kunzisoft.keepass.database.PwGroup;

public class CopyEntryRunnable extends ActionNodeDatabaseRunnable {

	private static final String TAG = CopyEntryRunnable.class.getName();

	private PwEntry mEntryToCopy;
	private PwEntry mEntryCopied;
	private PwGroup mNewParent;

	public CopyEntryRunnable(Context context, Database db, PwEntry oldE, PwGroup newParent, AfterActionNodeOnFinish afterAddNode) {
		this(context, db, oldE, newParent, afterAddNode, false);
	}

	public CopyEntryRunnable(Context context, Database db, PwEntry oldE, PwGroup newParent, AfterActionNodeOnFinish afterAddNode, boolean dontSave) {
		super(context, db, afterAddNode, dontSave);

		this.mEntryToCopy = oldE;
		this.mNewParent = newParent;
	}

	@Override
	public void run() {
		// Update entry with new values
        mEntryCopied = mDb.copyEntry(mEntryToCopy, mNewParent);

        if (mEntryCopied != null) {
            mEntryCopied.touch(true, true);

            // Commit to disk
            super.run();
        } else {
            Log.e(TAG, "Unable to create a copy of the entry");
        }
	}

	@Override
	protected void onFinish(boolean success, String message) {
		if ( !success ) {
			// If we fail to save, try to delete the copy
			try {
				mDb.deleteEntry(mEntryCopied);
			} catch (Exception e) {
				Log.i(TAG, "Unable to delete the copied entry");
			}
		}
        callbackNodeAction(success, message, mEntryToCopy, mEntryCopied);
	}
}

File: app/src/main/java/com/kunzisoft/keepass/database/action/node/CopyEntryRunnable.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.action.node

import androidx.fragment.app.FragmentActivity
import android.util.Log
import com.kunzisoft.keepass.R
import com.kunzisoft.keepass.database.element.Database
import com.kunzisoft.keepass.database.element.EntryVersioned
import com.kunzisoft.keepass.database.element.GroupVersioned

class CopyEntryRunnable constructor(
        context: FragmentActivity,
        database: Database,
        private val mEntryToCopy: EntryVersioned,
        private val mNewParent: GroupVersioned,
        afterAddNodeRunnable: AfterActionNodeFinishRunnable?,
        save: Boolean)
    : ActionNodeDatabaseRunnable(context, database, afterAddNodeRunnable, save) {

    private var mEntryCopied: EntryVersioned? = null

    override fun nodeAction() {
        // Condition
        var conditionAccepted = true
        if(mNewParent == database.rootGroup && !database.rootCanContainsEntry())
            conditionAccepted = false
        if (conditionAccepted) {
            // Update entry with new values
            mNewParent.touch(modified = false, touchParents = true)
            mEntryCopied = database.copyEntryTo(mEntryToCopy, mNewParent)
        } else {
            // Only finish thread
            throw Exception(context.getString(R.string.error_copy_entry_here))
        }

        mEntryCopied?.apply {
            touch(modified = true, touchParents = true)
        } ?: Log.e(TAG, "Unable to create a copy of the entry")
    }

    override fun nodeFinish(result: Result): ActionNodeValues {
        if (!result.isSuccess) {
            // If we fail to save, try to delete the copy
            try {
                mEntryCopied?.let {
                    database.deleteEntry(it)
                }
            } catch (e: Exception) {
                Log.i(TAG, "Unable to delete the copied entry")
            }
        }
        return ActionNodeValues(result, mEntryToCopy, mEntryCopied)
    }

    companion object {
        private val TAG = CopyEntryRunnable::class.java.name
    }
}
